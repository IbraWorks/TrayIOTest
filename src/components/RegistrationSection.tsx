import React from "react";
import { IRegistrationValues } from "../interfaces/IFormValues";
import { Form, Button } from "semantic-ui-react";
import { useForm } from "react-hook-form";
interface IProps {
  handleData: (data: IRegistrationValues) => void;
  registrationValues: IRegistrationValues;
}

const RegistrationSection: React.FC<IProps> = ({ handleData, registrationValues }) => {
  const { register, handleSubmit, errors } = useForm();

  const onSubmit = (data: any) => {
    handleData(data);
  };

  return (
    <div>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Field required>
          <label style={{ float: "left" }}>Name</label>
          <input
            name="name"
            placeholder="John Smith"
            defaultValue={registrationValues.name}
            ref={register({ required: true })}
          />
          {errors.name && errors.name.type === "required" && (
            <p style={{ color: "red" }}>Your name is required to register</p>
          )}
        </Form.Field>
        <Form.Field>
          <label style={{ float: "left" }}>Role</label>
          <input
            name="role"
            placeholder="Software Engineer"
            defaultValue={registrationValues.role}
            ref={register({ required: false })}
          />
        </Form.Field>
        <Form.Field required>
          <label style={{ float: "left" }}>Email</label>
          <input

            name="email"
            placeholder="email@example.com"
            defaultValue={registrationValues.email}
            ref={register({ required: true, pattern: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/ })}
          />
          {errors.email && errors.email.type === "required" && (
            <p style={{ color: "red" }}>Your email is required to register</p>
          )}
          {errors.email && errors.email.type === "pattern" && (
            <p style={{ color: "red" }}>Please enter a valid email address</p>
          )}
        </Form.Field>
        <Form.Field required>
          <label style={{ float: "left" }}>Password</label>
          <input
            name="password"
            type="password"
            placeholder="password"
            defaultValue={registrationValues.password}
            ref={register({
              required: true,
              minLength: 10,
              pattern: /(?=.*\d)(?=.*[a-z])(?=.*[A-Z])/,
            })}
          />
          {errors.password && errors.password.type === "required" && (
            <p style={{ color: "red" }}>A valid password is required to register</p>
          )}
          {errors.password && errors.password.type === "minLength" && (
            <p style={{ color: "red" }}>The minimum length is 10</p>
          )}
          {errors.password && errors.password.type === "pattern" && (
            <p style={{ color: "red" }}>
              Password must contain at least one uppercase, at least one
              lowercase, and at least one digit
            </p>
          )}
        </Form.Field>
        <Button color="green" style={{ float: "right" }} type="submit">
          Submit
        </Button>
      </Form>
    </div>
  );
};

export default RegistrationSection;
