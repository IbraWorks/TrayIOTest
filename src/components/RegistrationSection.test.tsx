import { shallow } from 'enzyme'
import React from 'react'
import RegistrationSection from './RegistrationSection';

describe("RegistrationSection", () => {
    

    it('should render', () => {
        const mockHandleData = jest.fn();
        const registrationValues = {
            name: "",
            role: "",
            email: "",
            password: ""
          }
        const wrapper = shallow(<RegistrationSection handleData={mockHandleData} registrationValues={registrationValues}/>)
        expect(wrapper.exists()).toBe(true)
    });

    it('matches snapshot', () => {
        const mockHandleData = jest.fn();
        const registrationValues = {
            name: "",
            role: "",
            email: "",
            password: ""
          }
        const wrapper = shallow(<RegistrationSection handleData={mockHandleData} registrationValues={registrationValues}/>)
        expect(wrapper.html()).toMatchSnapshot();
    });


    it('submitting the form should call handle data', () => {
        const mockHandleData = jest.fn();
        const registrationValues = {
            name: "John Smith",
            role: "Software Engineer",
            email: "email@example.com",
            password: "Password12345"
          }
        const wrapper = shallow(<RegistrationSection handleData={mockHandleData} registrationValues={registrationValues}/>)
        wrapper.find('Form').simulate('submit');
        expect(mockHandleData).toBeCalledTimes(1);
    });

    it('has inputs for name, role, email and password', () => {
        const mockHandleData = jest.fn();
        const registrationValues = {
            name: "",
            role: "",
            email: "",
            password: ""
          }
        const wrapper = shallow(<RegistrationSection handleData={mockHandleData} registrationValues={registrationValues}/>)
        expect(wrapper.find("input[name='name']").length).toEqual(1);
        expect(wrapper.find("input[name='role']").length).toEqual(1);
        expect(wrapper.find("input[name='email']").length).toEqual(1);
        expect(wrapper.find("input[name='password']").length).toEqual(1);
    });

    it('has a submit button', () => {
        const mockHandleData = jest.fn();
        const registrationValues = {
            name: "",
            role: "",
            email: "",
            password: ""
          }
        const wrapper = shallow(<RegistrationSection handleData={mockHandleData} registrationValues={registrationValues}/>)
        expect(wrapper.find("Button").length).toEqual(1);
        expect(wrapper.find('Button').props()).toEqual({
            "as": "button",
            "children": "Submit",
            "color": "green",
            "style": {float: "right"},
            "type": "submit",
        })
    });
})