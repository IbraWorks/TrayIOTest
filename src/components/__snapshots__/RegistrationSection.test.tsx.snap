// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationSection matches snapshot 1`] = `"<div><form class=\\"ui form\\"><div class=\\"required field\\"><label style=\\"float:left\\">Name</label><input name=\\"name\\" placeholder=\\"John Smith\\" value=\\"\\"/></div><div class=\\"field\\"><label style=\\"float:left\\">Role</label><input name=\\"role\\" placeholder=\\"Software Engineer\\" value=\\"\\"/></div><div class=\\"required field\\"><label style=\\"float:left\\">Email</label><input name=\\"email\\" placeholder=\\"email@example.com\\" value=\\"\\"/></div><div class=\\"required field\\"><label style=\\"float:left\\">Password</label><input type=\\"password\\" name=\\"password\\" placeholder=\\"password\\" value=\\"\\"/></div><button style=\\"float:right\\" type=\\"submit\\" class=\\"ui green button\\">Submit</button></form></div>"`;
